tpls/desul/include/desul/atomics/Fetch_Op_OpenACC.hpp

```diff
@@ -0,0 +1,491 @@
+#pragma acc routine seq
+template <class T>
+acc_enable_if_supported_arithmetic_type<T,T> device_atomic_fetch_add(
+T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { tmp = ptr[0]; ptr[0] += val; }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+acc_enable_if_supported_arithmetic_type<T,T> device_atomic_fetch_inc(
+T* ptr, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { tmp = ptr[0]; ptr[0] += T(1); }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+acc_enable_if_supported_arithmetic_type<T,T> device_atomic_fetch_sub(
+T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { tmp = *ptr; *ptr -= val; }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+acc_enable_if_supported_arithmetic_type<T,T> device_atomic_fetch_dec(
+T* ptr, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { tmp = *ptr; *ptr -= T(1); }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+acc_enable_if_supported_arithmetic_type<T,T> device_atomic_fetch_mul(
+T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { tmp = *ptr; *ptr *= val; }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+acc_enable_if_supported_arithmetic_type<T,T> device_atomic_fetch_div(
+T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { tmp = *ptr; *ptr /= val; }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+acc_enable_if_supported_arithmetic_type<T,T> device_atomic_fetch_lshift(
+T* ptr, const unsigned int val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { tmp = *ptr; *ptr = *ptr << val; }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+acc_enable_if_supported_arithmetic_type<T,T> device_atomic_fetch_rshift(
+T* ptr, const unsigned int val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+#pragma acc atomic capture
+  { tmp = *ptr; *ptr = *ptr >> val; }
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+T device_atomic_fetch_mod(
+T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+  if(acc_on_device(acc_device_not_host)) {
+    printf("Kokkos Error in device_atomic_fetch_mod(): Not supported atomic "
+                  "operation in the OpenACC backend\n");
+  }
+  tmp = *ptr; *ptr = *ptr % val;
+  return tmp;
+}
+
+#pragma acc routine seq
+template <class T>
+T device_atomic_fetch_max(
+T* ptr, const T val, MemoryOrderRelaxed, MemoryScopeDevice) {
+  T tmp;
+  if(acc_on_device(acc_device_not_host)) {
+    printf("Kokkos Error in device_atomic_fetch_max(): Not supported atomic "
+                  "operation in the OpenACC backend\n");
+  }
+  tmp = *ptr; *ptr = std::max(*ptr, val);
+  return tmp;
+}
+
+#ifdef KOKKOS_COMPILER_NVHPC
```

Comment:
Did you look at the assembly?  Is there any actual benefit from specializing for these rather than doing a regular acc atomic capture?

Reply:
In this case, the specialization is not for the performance benefits. The OpenACC C/C++ standard does not support atomic min/max operations, but if targeting NVIDIA GPUs with NVHPC, we can use CUDA atomic min/max intrinsics.
This specialization is an NVHPC-specific extension.